cellarium/ml/layers/transformer.py

```diff
@@ -0,0 +1,196 @@
+        attention_mask_ncc: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Args:
+            hidden_state_ncd:
+                Hidden state tensor of shape ``(n, c, d)``.
+            attention_mask_ncc:
+                Attention mask tensor of shape ``(n, c, c)``.
+
+        Returns:
+            The output hidden state tensor of shape ``(n, c, d)``.
+        """
+        hidden_state_ncd = self.normadd1(hidden_state_ncd, lambda X: self.attention(X, X, X, attention_mask_ncc))
+        return self.normadd2(hidden_state_ncd, lambda Y: self.ffn(Y))  # _ncd
+
+
+class Transformer(nn.Module):
+    """
+    Transformer model.
+
+    Args:
+        d_model:
+            Dimensionality of the embeddings and hidden states.
+        d_ffn:
+            Dimensionality of the inner feed-forward layers.
+        use_bias:
+            Whether to use bias in the linear transformations and norm-add layers.
+        n_heads:
+            Number of attention heads.
+        n_blocks:
+            Number of transformer blocks.
+        dropout_p:
+            Dropout probability.
+        attention_logits_scale:
+            Multiplier for the attention scores.
+        attention_backend:
+            Backend for the attention computation.
+        Wqkv_initializer:
+            Initializer for the query, key, and value linear transformations.
+        Wo_initializer:
+            Initializer for the output linear transformation.
+        dense1_initializer:
+            Initializer for the first dense layer.
+        dense2_initializer:
+            Initializer for the second dense layer.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        d_ffn: int,
+        use_bias: bool,
+        n_heads: int,
+        n_blocks: int,
+        dropout_p: float,
+        attention_logits_scale: float,
+        attention_backend: Literal["math", "flash", "mem_efficient", "torch"],
+        attention_softmax_fp32: bool,
+        Wqkv_initializer: dict[str, Any],
+        Wo_initializer: dict[str, Any],
+        dense1_initializer: dict[str, Any],
+        dense2_initializer: dict[str, Any],
+    ) -> None:
+        super().__init__()
+        self.n_blocks = n_blocks
+        self.blocks = nn.ModuleList(
+            [
+                TransformerBlock(
+                    d_model,
+                    d_ffn,
+                    use_bias,
+                    n_heads,
+                    dropout_p,
+                    attention_logits_scale,
+                    attention_backend,
+                    attention_softmax_fp32,
+                    Wqkv_initializer,
+                    Wo_initializer,
+                    dense1_initializer,
+                    dense2_initializer,
+                )
+                for _ in range(n_blocks)
+            ]
+        )
+        self.ln = nn.LayerNorm(d_model, use_bias)
+
+        self._reset_parameters()
+
+    def _reset_parameters(self) -> None:
+        self.ln.reset_parameters()
```

Comment:
The idea behind `_reset_parameters` methods is that you can call `self.apply(lambda x: x._reset_parameters())` from the top model (e.g., `CellariumGPT`) which will recursively call `_reset_parameters` on all children.

Here `Transformer._reset_parameters` only has to reset `LayerNorm` parameters because MHA and FFN layers already have their own `_reset_parameters` methods defined.